python/singa/model.py

```diff
@@ -0,0 +1,229 @@
+
+                self.sgd = opt.SGD(lr=0.01)
+
+            def forward(self, x):
+                y = self.conv1(x)
+                y = self.conv2(y)
+                return y
+
+            def train_one_batch(self, x, y):
+                out = self.forward(x)
+                loss = autograd.softmax_cross_entropy(out, y)
+                self.sgd.backward_and_update(loss)
+                return out, loss
+
+    """
+
+    def __init__(self):
+        """
+        Initializes internal Model state
+        """
+        self.training = True
+        self.buffered = False
+        self.graph_mode = True
+        self.sequential = False
+        self.initialized = False
+        self._device = get_default_device()
+
+        self._results = None
+
+    def compile(self, inputs, is_train=True, use_graph=False, sequential=False):
+        self._device.EnableGraph(True)
+        self.forward(*inputs)
+        self._device.EnableGraph(False)
+        self._device.ResetGraph()
+        autograd.training = is_train
+        self.training = is_train
+        self.graph_mode = use_graph
+        self.sequential = sequential
+
+    def forward(self, *input):
+        """Defines the computation performed at every call.
+
+        Should be overridden by all subclasses.
+
+        Args:
+            *input: the input training data for the model
+
+        Returns:
+            out: the outputs of the forward propagation.
+        """
+        raise NotImplementedError
+
+    def train_one_batch(self, *input):
+        raise NotImplementedError
+
+    def train(self, mode=True):
+        """Set the model in evaluation mode.
+
+        Args:
+            mode(bool): when mode is True, this model will enter training mode
+        """
+        self.training = mode
+        autograd.training = mode
+
+    def eval(self):
+        """Sets the model in evaluation mode.
+        """
+        self.train(mode=False)
+        autograd.training = False
+
+    def graph(self, mode=True, sequential=False):
+        """ Turn on the computational graph. Specify execution mode.
+
+        Args:
+            mode(bool): when mode is True, model will use computational graph
+            sequential(bool): when sequential is True, model will execute ops
+            in the graph follow the order of joining the graph
+        """
+        self.graph_mode = mode
+        self.sequential = sequential
+
+    def on_device(self, device):
+        """Sets the target device.
+
+        The following training will be performed on that device.
+
+        Args:
+            device(Device): the target device
+        """
+        self._device = device
+
+    def __get_name__(self):
+        return self.__class__.__name__
+
+    def __call__(self, *input, **kwargs):
+        if self.training:
+            return self.train_one_batch(*input, **kwargs)
+        else:
+            return self.forward(*input, **kwargs)
+
+    def get_params(self):
+        params = super().get_params()
```

Comment:
move it into layer.py as a method of Layer?
A layer may have sublayers.
Then Model will inherit this method.

Reply:
Yes, it is a method of layer. `model` is kind of overloading it as `Layer.get_params()` returns nested dictionary, and `model.get_params()` flatten the nested dictionary